aten/src/ATen/native/cuda/AtomicOps.cuh

```diff
@@ -0,0 +1,344 @@
+
```

Comment:
So now we have THCAtomics.cuh and this file, which duplicates THCAtomic.cuh functionality/implementation for add. This is pretty confusing. 
Given that we only dispatch floating types, implementing multiple ops for all integer variants seems excessive? What's needed is atomicMul for bfloat16, half, float and double, right, and that can probably go to THCAtomics.

Reply:
I thought I'd future proof the implementation by implementing for all the data types. I've just added CUDA-specific tests using C++ for all the data types.

IMO the excessive bloat caused in the THCAtomics file justifies putting this in a separate file and changing all the usages of THCAtomics `gpuAtomicAdd` to use the new implementation in a later PR.

Reply:
Not if we limit the extensions to only the necessary functions. 
There are a few problems with switching things to the new implementations:
1) There are extensions that depend on THCAtomics, changing them for no good reason is tedious
2) (fixable) functions here are `void` instead of returning old value, and some extensions rely on returning old values. 
3) there's fastAtomicAdd implementation in kernelUtils.cuh, which is immensely better for halfs, if we want to go through the trouble of switching all atomic uses, we might as well switch to it
4) the integer implementations are very slow, and only cause user confusion (#42109), there's no point having them just for the sake of having them
In short, I think we should be adding only functions and specializations we really need, and we better keep them in the same file, so that it's easier to keep track of what we have and what we don't.

Reply:
Ok makes sense. I'll revert to adding the multiplication to the old file and use the old functions for addition.